File: solutions/haskell/152-combination-sum-iv.hs
{-
  Problem 152: Combination Sum IV (LeetCode 377)
  Difficulty: Med
  Language: Haskell
-}
combinationSum4 :: [Int] -> Int -> Int
combinationSum4 nums target =
  let dp = foldl (\dp i ->
        dp ++ [sum [dp !! (i - n) | n <- nums, n <= i]]
      ) [1] [1..target]
  in dp !! target

File: solutions/ocaml/152-combination-sum-iv.ml
(*
  Problem 152: Combination Sum IV (LeetCode 377)
  Difficulty: Med
  Language: OCaml
*)
let combination_sum4 nums target =
  let dp = Array.make (target + 1) 0 in
  dp.(0) <- 1;
  for i = 1 to target do
    dp.(i) <- List.fold_left (fun sum n ->
      if n <= i then sum + dp.(i - n) else sum) 0 nums
  done; dp.(target)
